libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1006,6 +1008,152 @@ std::string YulUtilFunctions::resizeDynamicArrayFunction(ArrayType const& _type)
+					let arrayDataStart := <dataPosition>(array)
+					let deleteStart := add(arrayDataStart, div(add(newLen, 31), 32))
+
+					// we have to partially clear last slot that is still used
+					let offset := mod(newLen, 32)
+					if gt(offset, 0) { <partialClearStorageSlot>(sub(deleteStart, 1), offset) }
+
+					<clearStorageRange>(deleteStart, add(arrayDataStart, div(add(oldLen, 31), 32)))
+
+					sstore(array, or(<shl>(1, newLen), 1))
+				}
+				default {
+					switch gt(oldLen, 31)
+					case 0 {
+						let arrayDataStart := <dataPosition>(array)
+						// clear whole old array, as we are transforming to short bytes array
+						<clearStorageRange>(arrayDataStart, add(arrayDataStart, div(add(oldLen, 31), 32)))
+						<transitLongToShort>(array, newLen)
+					}
+					default {
+						<changeLengthShort>(array, sload(array), newLen)
+					}
+				}
+			})")
+			("functionName", functionName)
+			("dataPosition", arrayDataAreaFunction(_type))
+			("partialClearStorageSlot", partialClearStorageSlotFunction())
+			("clearStorageRange", clearStorageRangeFunction(*_type.baseType()))
+			("shl", shiftLeftFunctionDynamic())
+			("transitLongToShort", byteArrayTransitLongToShortFunction(_type))
+			("changeLengthShort", changeLengthOfShortByteArrayFunction(_type))
+			.render();
+	});
+}
+
+string YulUtilFunctions::increaseByteArraySizeFunction(ArrayType const& _type)
+{
+	string functionName = "byte_array_increase_size_to_unpacked_" + _type.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(array, oldLen, newLen) {
+				switch lt(oldLen, 32)
+				case 0 {
+					// in this case array stays unpacked, so we just set new length
+					sstore(array, or(<shl>(1, newLen), 1))
+				}
+				default {
+					let data := and(not(0xff), sload(array))
+					switch lt(newLen, 32)
+					case 0 {
+						// we need to copy elements to data area as we changed array from packed to unpacked
+						sstore(<dataPosition>(array), data)
+						sstore(array, or(<shl>(1, newLen), 1))
+					}
+					default {
+						// here array stays packed, we just need to increase length
+						sstore(array, or(<shl>(1, newLen), data))
+					}
+				}
+			})")
+			("functionName", functionName)
+			("dataPosition", arrayDataAreaFunction(_type))
+			("shl", shiftLeftFunctionDynamic())
+			.render();
+	});
+}
+
+string YulUtilFunctions::byteArrayTransitLongToShortFunction(ArrayType const& _type)
+{
+	string functionName = "transit_byte_array_long_to_short_" + _type.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(array, len) {
+				// we need to copy elements from old array to new
+				// we want to copy only elements that are part of the array after resizing
+				let mask := <shl>(mul(8, sub(32, len)), <ones>)
+				let dataPos := <dataPosition>(array)
+				let data := and(sload(dataPos), mask)
+				sstore(array, or(data, <shl>(1, len)))
```

Comment:
Please try to avoid dynamic shift

Reply:
I am pretty sure that I cant avoid it here `<shl>(mul(8, sub(32, len)), <ones>)`